pkg/controller/mssql/grant/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	}, nil
+}
+
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	permissions := strings.Join(cr.Spec.ForProvider.Permissions.ToStringSlice(), ", ")
+
+	query := fmt.Sprintf("GRANT %s TO %s", permissions, mssql.QuoteIdentifier(username))
+	return managed.ExternalCreation{}, errors.Wrap(c.db.Exec(ctx, xsql.Query{String: query}), errGrant)
+}
+
+func (c *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotGrant)
+	}
+
+	desired := cr.Spec.ForProvider.Permissions.ToStringSlice()
+	observed, err := c.getPermissions(ctx, *cr.Spec.ForProvider.User)
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+	toGrant, toRevoke := diffPermissions(desired, observed)
+
+	if len(toRevoke) > 0 {
+		query := fmt.Sprintf("REVOKE %s FROM %s",
+			strings.Join(toRevoke, ", "), mssql.QuoteIdentifier(*cr.Spec.ForProvider.User))
+		if err = c.db.Exec(ctx, xsql.Query{String: query}); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, errRevoke)
+		}
+	}
+	if len(toGrant) > 0 {
+		query := fmt.Sprintf("GRANT %s TO %s",
+			strings.Join(toGrant, ", "), mssql.QuoteIdentifier(*cr.Spec.ForProvider.User))
+		if err = c.db.Exec(ctx, xsql.Query{String: query}); err != nil {
+			return managed.ExternalUpdate{}, errors.Wrap(err, errGrant)
+		}
+	}
+	return managed.ExternalUpdate{}, nil
+}
+
+func (c *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+
+	query := fmt.Sprintf("REVOKE %s FROM %s",
+		strings.Join(cr.Spec.ForProvider.Permissions.ToStringSlice(), ", "),
+		mssql.QuoteIdentifier(username),
+	)
+	return errors.Wrap(c.db.Exec(ctx, xsql.Query{String: query}), errRevoke)
+}
+
+func stringValue(p *string) string {
+	if p == nil {
+		return ""
+	}
+	return *p
+}
+
+func (c *external) getPermissions(ctx context.Context, username string) ([]string, error) {
+	query := fmt.Sprintf(`SELECT pe.permission_name
+	FROM sys.database_principals AS pr  
+	JOIN sys.database_permissions AS pe  
+	    ON pe.grantee_principal_id = pr.principal_id  
+	WHERE
+	  pr.name = %s`, mssql.QuoteValue(username))
+	rows, err := c.db.Query(ctx, xsql.Query{String: query})
+	if err != nil {
+		return nil, errors.Wrap(err, errCannotGetGrants)
+	}
+	defer rows.Close() //nolint:errcheck
+
+	var permissions []string
+	for rows.Next() {
+		var grant string
+		if err := rows.Scan(&grant); err != nil {
+			return nil, errors.Wrap(err, errCannotGetGrants)
+		}
+		if strings.EqualFold(grant, "CONNECT") {
+			// Todo(turkenh): CONNECT permission is granted by default at user
```

Comment:
I am not happy with the current implementation and agree that late-init would be more explicit. 
The reason that I didn't choose that is related to how we name this kind, `Grant`. I believe this is a wrong choice here because we `Grant` or `Revoke` based on the permission list that users define in the CR. Ideally, I would prefer it something like `UserPermissions` or `DatabaseUserPermissions` but didn't go in that route to be consistent with others.

Also with the late-init option, not sure if it is the right action to bring all existing permissions and add to the desired list that user-provided, which might not always be the intention.

Reply:
We just had a quick discussion with @muvaf and he mentioned that they had a similar case in another resource and chose to just apply whatever user adds to the list. So, I'll remove the exception on `CONNECT` and include it in the example to show that it should/could be added in the list.